src/ray/object_manager/push_manager.h

```diff
@@ -0,0 +1,118 @@
+        chunk_send_fn(chunk_send_fn),
+        next_chunk_id(0),
+        chunks_remaining(num_chunks) {}
+};
+
+/// Manages rate limiting and deduplication of outbound object pushes.
+class PushManager {
+ public:
+  /// Create a push manager.
+  ///
+  /// \param max_chunks_in_flight Max number of chunks allowed to be in flight
+  ///                             from this PushManager (this raylet).
+  PushManager(int64_t max_chunks_in_flight)
+      : max_chunks_in_flight_(max_chunks_in_flight) {
+    RAY_CHECK(max_chunks_in_flight_ > 0) << max_chunks_in_flight_;
+  };
+
+  /// Start pushing an object subject to max chunks in flight limit.
+  ///
+  /// Duplicate concurrent pushes to the same destination will be suppressed.
+  ///
+  /// \param dest_id The node to send to.
+  /// \param obj_id The object to send.
+  /// \param num_chunks The total number of chunks to send.
+  /// \param send_chunk_fn This function will be called with args 0...{num_chunks-1}.
+  ///                      The caller promises to call PushManager::OnChunkComplete()
+  ///                      once a call to send_chunk_fn finishes.
+  void StartPush(const NodeID &dest_id, const ObjectID &obj_id, int64_t num_chunks,
+                 std::function<void(int64_t)> send_chunk_fn);
+
+  /// Called every time a chunk completes to trigger additional sends.
+  /// TODO(ekl) maybe we should cancel the entire push on error.
+  void OnChunkComplete(const NodeID &dest_id, const ObjectID &obj_id);
+
+  /// Return the number of chunks currently in flight. For testing only.
+  int64_t NumChunksInFlight() const { return chunks_in_flight_; };
+
+  /// Return the number of chunks remaining. For testing only.
+  int64_t NumChunksRemaining() const {
+    int total = 0;
+    for (const auto &pair : push_info_) {
+      total += pair.second->chunks_remaining;
+    }
+    return total;
+  }
+
+  /// Return the number of pushes currently in flight. For testing only.
+  int64_t NumPushesInFlight() const { return push_info_.size(); };
+
+  std::string DebugString() const {
+    std::stringstream result;
+    result << "PushManager:";
+    result << "\n- num pushes in flight: " << NumPushesInFlight();
+    result << "\n- num chunks in flight: " << NumChunksInFlight();
+    result << "\n- num chunks remaining: " << NumChunksRemaining();
+    result << "\n- max chunks allowed: " << max_chunks_in_flight_;
+    return result.str();
+  }
+
+ private:
+  /// Called on completion events to trigger additional pushes.
+  void ScheduleRemainingPushes();
+
+  /// Pair of (destination, object_id).
+  typedef std::pair<NodeID, ObjectID> PushID;
+
+  /// Max number of chunks in flight allowed.
+  const int64_t max_chunks_in_flight_;
+
+  /// Running count of chunks in flight, used to limit progress of in_flight_pushes_.
+  int64_t chunks_in_flight_ = 0;
+
+  /// Tracks all pushes with chunk transfers in flight.
+  absl::flat_hash_map<PushID, std::unique_ptr<PushState>> push_info_;
```

Comment:
Why unique_ptr?

Reply:
I didn't want to run the risk of treating PushState as a value-type (unless we have DISALLOW_COPY_AND_ASSIGN?), and it's not perf sensitive, and had some issues with `map[key] = value` without defining a copy constructor. Open to suggestions on how to do it as a value type.

Reply:
Ah I think you just need to avoid accessing with `map[key]` and use `map.find(key)` and `map.emplace` to look up/insert instead. You could delete the copy constructor to prevent copies, but it's probably okay without it too.